src/python/pants/backend/python/util_rules/faas.py

```diff
@@ -335,21 +324,107 @@ async def digest_complete_platforms(
 
 
 @dataclass(frozen=True)
-class KnownRuntimeCompletePlatformRequest:
-    module: str
-    file_name: None | str
+class RuntimePlatformsRequest:
+    address: Address
+    target_name: str
+
+    runtime: PythonFaaSRuntimeField
+    complete_platforms: PythonFaaSCompletePlatforms
+
+
+@dataclass(frozen=True)
+class RuntimePlatforms:
+    interpreter_version: None | tuple[int, int]
+    pex_platforms: PexPlatforms = PexPlatforms()
+    complete_platforms: CompletePlatforms = CompletePlatforms()
+
+
+async def _infer_from_ics(
+    request: RuntimePlatformsRequest, python_setup: PythonSetup
+) -> tuple[int, int]:
+    ics = await Get(InterpreterConstraints, InterpreterConstraintsRequest([request.address]))
+
+    # (slight) future proofing: use the overall Pants interpreter universe, rather than the FaaS
```

Comment:
Either I'm reading this incorrectly, or isn't this the opposite of future-proofing? You could upgrade Pants and suddenly break the build because your ICs now support a new Python version that Pants knows about, even though AWS has not added support for that version.

Reply:
Ah hm... yes, that's the same sort of problem as pants updating its list of known runtimes for each FaaS service. 

Some options:

1. Switch to using the known runtimes for each FaaS platform, eg `>=3.10` would work for Lambda (since pants only knows up to 3.10) now, but become ambiguous once Lambda adds support for 3.11 and pants learns about it
2. Keep as is, using pants' global universe, so `>=3.11` is unambiguous now but becomes ambiguous once pants learns about 3.12 (but, a user in this circumstance may find other surprises with the newly permitted 3.12, not just in their FaaS targets)
3. Use an even bigger universe (like 2.7, 3.0, 3.1, ..., 3.49, 3.50) to truly require the ICs are narrow

I'm still inclined to 2 (what's implemented), but weakly.

Thoughts?

Reply:
We could also be strict and just not infer at all if the ICs are compatible with more than one minor version. That is the most future proof.

Reply:
Okay, that sounds like 3: using a very wide universe. Unless there's other ways to extract the compatible minor versions from ICs without a universe?

I'll implement, thanks.

Reply:
Hmm, yeah, or we can just cheat and pattern-match on "==maj.min.*" and ">=maj.min,<maj.min+1". If you have more complex situations than that then you're probably asking for trouble anyway. For example, if you pin to a specific patch version, or exclude a patch version, we can't even know if that is compatible with the patch version the FAAS provider will run you on.

Reply:
Yeah, that's a good point.

I've done something along those lines, although I'm not exactly sure it's "just cheat"-ing, given the fiddliness of the code required. 😅